cdap-unit-test/src/test/java/co/cask/cdap/test/app/AppWithCustomTx.java

```diff
@@ -322,42 +351,193 @@ public void destroy() {
     }
   }
 
-  public static class TimeoutHandler extends AbstractHttpServiceHandler {
+  public static class TxHandler extends AbstractHttpServiceHandler {
+
+    @Override
+    public void initialize(HttpServiceContext context) throws Exception {
+      super.initialize(context);
+      recordTransaction(getContext(), HANDLER_TX, INITIALIZE);
+      attemptNestedTransaction(getContext(), HANDLER_TX, INITIALIZE_NEST);
+    }
 
-    // service context does not have Transactional, no need to test lifecycle methods
+    @Override
+    public void destroy() {
+      recordTransaction(getContext(), HANDLER_TX, DESTROY);
+      attemptNestedTransaction(getContext(), HANDLER_TX, DESTROY_NEST);
+      super.destroy();
+    }
 
     @PUT
-    @Path("test")
-    public HttpContentConsumer handle(HttpServiceRequest request, HttpServiceResponder responder) {
+    @Path("tx")
+    public HttpContentConsumer tx(HttpServiceRequest request, HttpServiceResponder responder) {
+      recordTransaction(getContext(), HANDLER_TX, RUNTIME);
+      attemptNestedTransaction(getContext(), HANDLER_TX, RUNTIME_NEST);
+
       return new HttpContentConsumer() {
 
+        String body = null;
+
         @Override
         public void onReceived(ByteBuffer chunk, Transactional transactional) throws Exception {
-          executeRecordTransaction(transactional, CONSUMER, RUNTIME_TX, TIMEOUT_CONSUMER_RUNTIME);
+          body = Bytes.toString(chunk);
+          if (FAIL_CONSUMER.equals(body)) {
+            LOG.warn("Failing consumer because the request was '" + body + "'");
+            throw new RuntimeException(body);
+          }
+          recordTransaction(getContext(), CONSUMER_TX, RUNTIME);
+          executeRecordTransaction(getContext(), CONSUMER_TX, RUNTIME_TX, TIMEOUT_CONSUMER_RUNTIME);
+          executeRecordTransaction(transactional, CONSUMER_TX, RUNTIME_TX_T, TIMEOUT_CONSUMER_RUNTIME);
+          executeAttemptNestedTransaction(getContext(), CONSUMER_TX, RUNTIME_NEST);
```

Comment:
here, because the method can use both the transactional and the service context, we want to test nested transactions with any combination of the two. They all should fail.